### ISSUE TITLE
`defaultdict` doesn't recognize value type from the `map` argument

### ISSUE THREAD

**Issue by Andrej730** [microsoft/pyright#8033]
It seems that `dd` should have a type `defaultdict[str, set[int]]` just as `d2` does have `dict[str, set[int]]` but `dd` type appears as `defaultdict[str, set[Unknown]]` instead.

```python
from typing import assert_type

s: str = "5"
i: int = 2
f: float = 2.55

d1 = {s: i}
reveal_type(d1) # dict[str, int]

d2 = {s: {i}}
reveal_type(d2) # dict[str, set[int]]

from collections import defaultdict
dd = defaultdict(set, {s: {i}})
# assert_type" mismatch: expected "defaultdict[str, set[int]]" but received "defaultdict[str, set[Unknown]]" 
assert_type(dd, defaultdict[str, set[int]])
```

[pyright-play](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCGAzkQKZ4D68CpAsAFANEBcURMIUAvFAEQCsvBklap8PAEwNgrYABswBcVAkA6fvwYMAJgEZuUAN4tMAXwYhSAN1IE5VRKQAUegJRQAxFG1IAxjABtdhAAGkwUGABdLXptCQNjVkMkU3N6Sxs7BxoXCXcvH38gjjCyQLFI6MZ6UEgoXzA5OVJ-JDAUIkxsXHxtUmACAFc5GEKYHW0DPoHh0b8YJzKwxKMU01cGL2IySmpSXigIJCIIJV8AC1ZSAA8af1JJ3mmhkbHi0LZScojKg4AjQb4Sy%2BUhIGyPZ6zN7BUpfAIAVRQAGsUGAAO4oX5QBjbcgwbLObTaMKQ17zd6w75RSIbehAA)

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

In the case of `d1` and `d2`, pyright is using its normal inference rules to infer the type of a dictionary expression. These inference rules are documented [here](https://microsoft.github.io/pyright/#/type-inference?id=dictionary-expressions).

In the case of `dd`, pyright is not using inference. Instead, it is following the standard evaluation rules for a constructor call. In particular, it's using the `__init__` overload:

```python
class defaultdict(dict[_KT, _VT]):
    ...
    @overload
    def __init__(
        self: defaultdict[str, _VT],  # pyright: ignore[reportInvalidTypeVarUse]  #11780
        default_factory: Callable[[], _VT] | None,
        map: SupportsKeysAndGetItem[str, _VT],
        /,
        **kwargs: _VT,
    ) -> None: ...
```

As a first argument (for `default_factor`), you're passing the argument `set`. This is converted to a `Callable` using the rules spelled out in [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/constructors.html#converting-a-constructor-to-callable). I won't go through all the details, but you can work it out yourself if you'd like to understand it in more detail. `_VT` gets the type `Unknown` as part of this process. 

To get the effect you're looking for, you need to change your code to:

```python
dd = defaultdict(set[int], {s: {i}})
```

**Comment by Andrej730** [comment]
If it bases the typing on the `default_factory` output type then shouldn't it throw an error initializing defaultdicts below?

Also, it seems to behave differently depending on the types - in the first case defaultdict type was extended by the mapping's value type but in the second case it wasn't.

```python
from typing import assert_type
from collections import defaultdict

s: str = "5"
i: int = 2
f: float = 2.55

# shouldn't error occur here?
# e.g. issue with argument SupportsKeysAndGetItem[_KT, _VT] where _VT = set[float]
# doesn't match the return type of the first argument Callable[[], _VT] where _VT = set[int]
dd = defaultdict(set[int], {s: {f}})
# "assert_type" mismatch: expected "defaultdict[str, set[int]]" but received "defaultdict[str, set[int] | set[float]]"
assert_type(dd, defaultdict[str, set[int]])

# also should be show an error?
dd0 = defaultdict(set[float], {s: {i}})
# though this works
assert_type(dd0, defaultdict[str, set[float]])
```
[pyright-play](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCGAzkQKZ4D68CpAsAFCiRQDGYANu6SzEmCkUzZc%2BACalgBAK7sYopDwYMiALihEYIKAF4oAIgCsehkjWp8ugEwNga4OzAELUSwDoDBpfQDE6gBZgMqIoAOT45OBaYCwsUlp%2B5KQA-Ay%2BpK5orpgkUqRQAO5IMH6EIGhSEKQo%2BADKUgg4eEQA0qRwRACCKKIA4qQwAJIwpBAA2hTNACoANFAUAGqTALoFCSB5C5M66v2j9o4wS6lQomCkRKH4EE4sJcV56zBxKFDASCAapeWV1VAAwgROAQAEZcUajJazTYrfJrDaLbZkGCjcxHeiiUTbcSSGRyBQwAAUSJR1UhUAA3qoKcAAL40gCUxz0xDIlGopD0UAgSCI1xgtzUpAAHjQeKRMXpsdJZPIeKMNCBZsTUUtOcCpPh1ixSEgAG7i-RS3Gy5EKpW7VFQAA%2BO2R%2BycS1VDBZ5BgVEQpAJGNmRpl%2BPlmnNyJVjPox0BRDA-kC7ExwLyRAC%2BUILwiuBS6NEAAYsRJpXieETdvbDrNKWpyUg6aHfMVAmg7n4eQVcABrIjOkiu900L3Zn1543%2Bs22vYOB1LelAA)

**Comment by erictraut** [comment]
Pyright is working correctly with your code sample above.

In the expression `defaultdict(set[int], {s: {f}})`, the constraint solver is able to find a solution for type variables `_KT` and `_VT` that satisfies all of the constraints in the call. The resulting type is `defaultdict[str, set[int] | set[float]]`. If you change your code to `defaultdict[str, int](set[int], {s: {f}})`, you will se an error because the additional constraint makes it impossible for the constraint solver to find a type for `_VT` that meet all of the constraints.

The expression `defaultdict(set[float], {s: {i}}]` gives a different answer because the subexpression `{s: {i}}` is inferred as `dict[str, set[float]]` by [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). If you change your code to the following, you will see the same evaluated type as the earlier example (`dd`) because bidirectional type inference cannot be used when you separate the expressions into two different statements.

```python
my_set = {s: {i}}
dd0 = defaultdict(set[float], my_set)
```

**Comment by Andrej730** [comment]
Thank you for the explanation, I've missed [pep484's the numerical tower part](https://peps.python.org/pep-0484/#the-numeric-tower) and this is why it was confusing for me when I noticed a `float` type treated differently.

The example `dd0 = defaultdict(set[float], my_set)` is also very interesting. 
Basically, bidirectional inference allowed to be more liberal and kind of join `int` and `float` to just `float`, but if type is already specified (`my_set` as `set[int]`) then constraint solver will always preserve the specified type and e.g. will never join it with `set[float]` to just `set[float]`?
